clinica/iotools/converters/genfi_to_bids/genfi_to_bids_utils.py

```diff
@@ -0,0 +1,650 @@
+            df["modality"].str.contains("fieldmap")
+        ]
+        .groupby(["source_id", "source_ses_id", "modality"])
+        .min()
+    )
+    df1 = df1.join(df2.rename(columns={"dir_num": "run_01_dir_num"}))
+    df_alt = df1.reset_index().assign(
+        fmap_type_p=lambda x: (x.run_01_dir_num != x.dir_num)
+    )
+    return df_alt.assign(
+        fmap=lambda x: x.fmap_type_p.apply(lambda y: "phasediff" if y else "magnitude")
+    )
+
+
+def merge_fieldmaps(df: DataFrame, df_fmap: DataFrame):
+    """Merges the dataframe containing the fieldmaps names
+
+    Parameters
+    ----------
+    df: Dataframe
+        Initial dataframe. Fieldmaps are identified on information extracted here.
+
+    df_fmap: Dataframe
+        Dataframe in which fieldmaps are identified.
+
+    Returns
+    -------
+    Dataframe
+        Dataframe containing the correct fieldmap for each fieldmap acquisition.
+    """
+    filter = ["source_id", "source_ses_id", "modality", "dir_num", "fmap"]
+    df1 = df.merge(df_fmap[filter], how="inner", on=filter[:-1])
+    df2 = df.merge(df_fmap[filter], how="left", on=filter[:-1])
+    df2 = df2[~df2["modality"].str.contains("fieldmap")]
+    df1 = df1.assign(suffix=lambda x: x.fmap)
+    return pd.concat([df2, df1], ignore_index=True)
+
+
+def compute_runs(df: DataFrame) -> DataFrame:
+    """This functions computes the run numbers.
+
+    Parameters
+    ----------
+    df: DataFrame
+        Dataframe without runs.
+
+    Returns
+    -------
+    DataFrame
+        Dataframe containing the correct run for each acquisition.
+    """
+    filter = ["source_id", "source_ses_id", "suffix", "dir_num"]
+    df1 = df[filter].groupby(filter).min()
+    df2 = df[filter].groupby(filter[:-1]).min()
+    df1 = df1.join(df2.rename(columns={"dir_num": "run_01_dir_num"}))
+    df_alt = df1.reset_index().assign(run=lambda x: (x.run_01_dir_num != x.dir_num))
+    return df_alt.assign(run_num=lambda df: df.run.apply(lambda x: f"run-0{int(x)+1}"))
+
+
+def complete_imaging_data(df_dicom: DataFrame) -> DataFrame:
+    """This function uses the raw information extracted from the images, to obtain all the information necessary for the BIDS
+
+    Parameters
+    ----------
+    df_dicom: DataFrame
+        Dataframe containing the data extracted from the images
+
+    Returns
+    -------
+    df_sub_ses_run: DataFrame
+        Dataframe with the data necessary for the BIDS
+    """
+
+    df_dicom = df_dicom.assign(
+        source_id=lambda df: df.source.apply(
+            lambda x: Path(Path(Path(x).parent).parent).name.split("-")[0]
+        )
+    )
+    df_dicom = df_dicom.assign(
+        source_ses_id=lambda df: df.source.apply(
+            lambda x: Path(Path(Path(x).parent).parent).name.split("-")[1]
+        )
+    )
+    df_dicom = df_dicom.assign(
+        origin=lambda df: df.source.apply(
+            lambda x: Path(Path(Path(Path(Path(x).parent).parent).parent).parent)
```

Comment:
It looks like this recursive pattern is present in various places in this PR. First, you should be able to simply call `parent` without casting to `Path` because the parent is already a `Path`, so it could be simplified to:

```python
lambda x: Path(x).parent.parent.parent.parent
```

It might be worth implementing a small function to make this more readable. Something like this maybe:

```python
def get_parent(path: str, n: int = 1) -> Path:
    """Get the path to the nth parent."""
    if n <= 0:
        return Path(path)
    return get_parent(Path(path).parent, n - 1)
```

These lambdas would then become:

```python
lambda x: get_parent(x, 2).name.split("-")[0]
lambda x: get_parent(x, 2).name.split("-")[1]
lambda x: get_parent(x, 4)
```

WDYT ?

Reply:
yup that works !